src/sinks/redis.rs

```diff
@@ -0,0 +1,596 @@
+    Some(event)
+}
+
+type RedisPipeResult = RedisResult<Vec<bool>>;
+
+impl Response for Vec<bool> {
+    fn is_successful(&self) -> bool {
+        self.iter().all(|x| *x)
+    }
+}
+
+#[derive(Debug, Clone)]
+struct RedisRetryLogic;
+
+impl RetryLogic for RedisRetryLogic {
+    type Error = RedisError;
+    type Response = Vec<bool>;
+
+    fn is_retriable_error(&self, _error: &Self::Error) -> bool {
+        true
+    }
+
+    fn should_retry_response(&self, response: &Self::Response) -> RetryAction {
+        if response.is_successful() {
+            return RetryAction::Successful;
+        }
+        RetryAction::Retry("Sent data to redis failed.".into())
+    }
+}
+
+#[derive(Clone)]
+pub struct RedisSink {
+    conn: ConnectionManager,
+    data_type: DataType,
+}
+
+impl Service<Vec<RedisKvEntry>> for RedisSink {
+    type Response = Vec<bool>;
+    type Error = RedisError;
+    type Future = BoxFuture<'static, RedisPipeResult>;
+
+    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, kvs: Vec<RedisKvEntry>) -> Self::Future {
+        let count = kvs.len();
+        let mut byte_size = 0;
+
+        let mut conn = self.conn.clone();
+        let mut pipe = redis::pipe();
+
+        for kv in kvs {
+            byte_size += kv.encoded_length();
+            match self.data_type {
+                DataType::List(method) => match method {
+                    Method::LPush => {
+                        if count > 1 {
+                            pipe.atomic().lpush(kv.key, kv.value);
+                        } else {
+                            pipe.lpush(kv.key, kv.value);
+                        }
+                    }
+                    Method::RPush => {
+                        if count > 1 {
+                            pipe.atomic().rpush(kv.key, kv.value);
+                        } else {
+                            pipe.rpush(kv.key, kv.value);
+                        }
+                    }
+                },
+                DataType::Channel => {
+                    if count > 1 {
+                        pipe.atomic().publish(kv.key, kv.value);
+                    } else {
+                        pipe.publish(kv.key, kv.value);
+                    }
+                }
+            }
+        }
+
+        Box::pin(async move {
+            let result: RedisPipeResult = pipe.query_async(&mut conn).await;
+            match result {
+                Ok(res) => {
+                    if res.is_successful() {
+                        emit!(RedisEventSent { count, byte_size });
+                    }
+                    Ok(res)
+                }
+                Err(error) => {
+                    emit!(RedisSendEventFailed {
+                        error: error.to_string()
+                    });
+                    Err(error)
+                }
+            }
```

Comment:
```suggestion
            match &result {
                Ok(res) if res.is_successful()  => emit!(RedisEventSent { count, byte_size })
                Err(error) => emit!(RedisSendEventFailed { error: error.to_string() })
            };
            result
```

Reply:
On a second thought: Maybe we should also log something in the `!res.is_successful` case?